packages/ai-jsx/src/batteries/docs.tsx

```diff
@@ -1,154 +1,503 @@
+     */
+    readonly name?: string,
+    /** The encoding used for this document. Defaults to utf-8. */
+    readonly encoding: string = 'utf-8',
+    /** The content type for this document. Defaults to plain text. */
+    readonly mimeType: string = 'text/plain'
+  ) {
+    this.pageContent = pageContent;
+    this.name = name;
+    this.encoding = encoding;
+    this.mimeType = mimeType;
+  }
+}
 
-  /**
-   * Metadata about the document, such as the URL at which it's hosted, the title, or when it was created.
-   */
-  metadata?: Metadata;
+/**
+ * A text document that can be used to ground responses.
+ */
+export class Document<DocumentMetadata extends Jsonifiable = Jsonifiable> {
+  constructor(
+    /**
+     * The content of the document. While this may often be a singleton, including separate
+     * meaningful pieces of text can aid in chunking/embedding.
+     */
+    readonly pageContent: string[],
+    /**
+     * A human-readable name for this document. This identifies the document to users and should
+     * be unique within a corpus. If the document is hosted on the web, its URL is a good choice.
+     */
+    readonly name?: string,
+    /**
+     * Other metadata about the document, such as when it was created.
+     */
+    readonly metadata?: DocumentMetadata
+  ) {
+    this.pageContent = pageContent;
+    this.name = name;
+    this.metadata = metadata;
+  }
 }
 
 /**
- * Chunk text into multiple documents.
- *
- * In LangChain, this is called "split". We use the phrase "chunk", because the cannonical term for the output
- * of this process is "chunks".
+ * A function that parses raw documents to produce text documents that can be used by an LLM.
  */
-export type Chunker = <Metadata extends Jsonifiable = Jsonifiable>(
-  doc: Document<Metadata>,
-  ...args: any[]
-) => Promise<Document<Metadata>[]>;
-export type ChunkMany = <Metadata extends Jsonifiable = Jsonifiable>(
-  doc: Document<Metadata>[],
-  ...args: any[]
-) => Promise<Document<Metadata>[]>;
+export type Parser<DocumentMetadata extends Jsonifiable = Jsonifiable> = (
+  raw: RawDocument
+) => Promise<Document<DocumentMetadata>>;
+
+function defaultParser<DocumentMetadata extends Jsonifiable = Jsonifiable>(
+  raw: RawDocument
+): Promise<Document<DocumentMetadata>> {
+  if (raw.mimeType.startsWith('text/')) {
+    const content = new TextDecoder(raw.encoding).decode(raw.pageContent);
+    return Promise.resolve(new Document([content], raw.name));
+    // TODO: Add support for other mime types.
+  }
+  throw new Error(`Unsupported mime type: ${raw.mimeType}`);
+}
 
 /**
- * Load Documents from a source.
+ * Objects related to loading a corpus. This is facilitated by a series of
+ * requests and corresponding responses.
+ *
+ * In addition to returning documents, each response may expand the corpus
+ * space in one or both of two dimensions:
+ *      Responses may include new partitions to be loaded. Partitions are
```

Comment:
Can we define `partition` before using it? I think I get from the context what it means, but I'm not totally sure. I think it would be helpful to define up front the relationship between documents, pages, partitions, etc.